lasagne/layers/recurrent.py

```diff
@@ -213,35 +230,88 @@ def __init__(self, incoming, input_to_hidden, hidden_to_hidden,
+        # precompute the input dot product
+        if (self.precompute_input and
+                input_to_hidden.output_shape[0] is not None and
+                input_shape[0] is not None and
+                input_shape[1] is not None and
+                (input_to_hidden.output_shape[0] !=
+                 input_shape[0]*input_shape[1])):
+            raise ValueError(
+                'When precompute_input == True, '
+                'input_to_hidden.output_shape[0] must equal '
+                'incoming.output_shape[0]*incoming.output_shape[1] '
+                '(i.e. batch_size*sequence_length) or be None but '
+                'input_to_hidden.output_shape = {} and '
+                'incoming.output_shape[0]*incoming.output_shape[1] = '
+                '{}'.format(input_to_hidden.output_shape[0],
+                            input_shape[0]*input_shape[1]))
+
+        # Check that the first dimension of input_to_hidden and
+        # hidden_to_hidden's outputs match when we won't precompute the input
+        # dot product
+        if (not self.precompute_input and
+                input_to_hidden.output_shape[0] is not None and
+                hidden_to_hidden.output_shape[0] is not None and
+                (input_to_hidden.output_shape[0] !=
+                 hidden_to_hidden.output_shape[0])):
+            raise ValueError(
+                'When precompute_input == False, '
+                'input_to_hidden.output_shape[0] must equal '
+                'hidden_to_hidden.output_shape[0] but '
+                'input_to_hidden.output_shape[0] = {} and '
+                'hidden_to_hidden.output_shape[0] = {}'.format(
+                    input_to_hidden.output_shape[0],
+                    hidden_to_hidden.output_shape[0]))
+
+        # Check that input_to_hidden and hidden_to_hidden output shapes match,
+        # but don't check a dimension of it's None for either shape
+        if not all(s1 is None or s2 is None or s1 == s2
+                   for s1, s2 in zip(input_to_hidden.output_shape[1:],
+                                     hidden_to_hidden.output_shape[1:])):
             raise ValueError("The output shape for input_to_hidden and "
-                             "input_to_hidden must be equal, but "
-                             "input_to_hidden.output_shape={} and "
-                             "hidden_to_hidden.output_shape={}".format(
+                             "hidden_to_hidden must be equal after the first "
+                             "dimension, but input_to_hidden.output_shape={} "
+                             "and hidden_to_hidden.output_shape={}".format(
                                  input_to_hidden.output_shape,
                                  hidden_to_hidden.output_shape))
 
+        # Check that input_to_hidden's output shape is the same as
+        # hidden_to_hidden's input shape but don't check a dimension of it's
```

Comment:
Haha, there were two places with this issue!  Hold on, let me fix this one too and rebase.